extension-manifest-v3/src/store/options.js

```diff
@@ -37,35 +49,66 @@ const Options = {
   },
   paused: [{ id: true, revokeAt: 0 }],
   installDate: '',
+  sync: true,
+  revision: 0,
   [store.connect]: {
     async get() {
       let { options = __PLATFORM__ !== 'safari' ? migrateFromMV2() : {} } =
         await chrome.storage.local.get(['options']);
 
-      // Migrate `trackerWheelDisabled` to `trackerWheel`
-      // INFO: `trackerWheel` option introduced in v9.7.0
-      if (options.trackerWheelDisabled !== undefined) {
-        options.trackerWheel = !options.trackerWheelDisabled;
-      }
-
       // Migrate `dnrRules` to `engines`
-      // INFO: `engines` option introduced in v10.1.0
+      // INFO: `engines` option introduced in v10.0.1
       if (options.dnrRules) {
         options.engines = options.dnrRules;
       }
 
-      // Set default value for keys, which type does no match the current one
-      Object.entries(options).forEach(([key, value]) => {
-        if (typeof value !== typeof Options[key]) {
-          delete options[key];
-          console.warn(`Saved options "${key}" key has wrong type, deleted`);
+      // Fetch options from the server for logged in users
+      if (options.terms && options.sync) {
+        const serverOptions = await getUserOptions();
+        if (serverOptions && serverOptions.revision > options.revision) {
+          for (const key of SYNC_OPTIONS) {
+            options[key] = serverOptions[key];
+          }
+          this.set(null, options, SYNC_OPTIONS.concat(['revision']));
         }
-      });
+      }
 
       return options;
     },
-    async set(_, options) {
-      if (options === null) options = {};
+    async set(_, options = {}, keys) {
+      if (!keys.includes('revision')) {
+        options = Object.assign({}, options, {
+          revision:
+            options.terms && options.sync && (await store.resolve(Session)).user
+              ? Date.now()
```

Comment:
Real time is notoriously difficult in distributed systems. An approach through counters is resilient to the situation that the local clock time is wrong.

The idea is roughly to keep a counter and increment it locally before sending. And when receiving (from the server), update it to the maximum of the local counter and the counter from the server. https://en.wikipedia.org/wiki/Lamport_timestamp

The algorithm needs also something to break ties. Not 100% sure if we need it (since the server ultimately keeps the state and can break ties by preferring the first received message). But if needed, the extension could create a random ID on startup before sending the first value. It doesn't need to be persisted; if the service worker dies, it can create a new random one.

Reply:
Thanks for your propositions. It looks like the counter thing might be tempting, but I think it is rather easy to find an edge case. 

Browser A (offline) updates multiple times, for example from 1 to 5.
Then Browser B (online) has an initial state of 1, and it updates once to 2.

The server after that events has state 2, with the newest value (correct one). Then browser A becomes online, so by the internal check (chrome.alarms), in the next hour fetches data from server, and.. it seems that local state 5 is newer than state 2 on the server, so it overwrites the server with values and state 5, which is wrong...

Reply:
Note that the mechanism would not change. That means collision detection and handle would be identical. In both cases, you need a timestamp to decide if an action took place before another. Using local system clock is a timestamp, but has a chance to introduce errors (if two devices are online at the same time, for instance). Idea like Lamport's timestamp are only another form of getting a timestamp.

But I think, in the scenario that you described, it is not avoidable to lose update (if they update the same fields): if two devices make changes without syncing with the server and (if merging is not possible) the only solution that I see is to decide which update came first and than reject the conflicting update later.

Let's look at the scenario and compare how clocks make a difference:

9:13, t=0: Browser A and Browser B started with a consistent state that they got from the server. Let's say assume it is 9:13 o'clock and the Lamport timestamp is 0.  

Browser A:
9:17, t=0: updates one field
9:17, t=1: (clock moved forward because of the change)
9:18, t=1: (the device is offline and unable to sync with the server)
9:23, t=1: updates another field
9:23, t=1: (clock doesn't move forward, only the internal dirty state changes)
... (more updates, but it always stays at t=1)

(in the meantime on Browser B)
10:17, t=0: updates one field
10:17, t=1: (clock moves forward)
10:18, t=1: able to sync with the server (the server now updates its database and the time progresses to t=2)
10:18, t=1: sees the timestamp from the server in the response (t=2).
10:18, t=2: (clock moves forward since the server clock is higher than the local clock)

(back to Browser A):
10:30, t=1: finally able to sync all changes with the server: the server sees that the changes are from the past. Now it is difficult because it is a collision: either it can be merged or not. If it can't be merged, then Browser A loses all updates and should get the latest state from the server. Or it can be merged.

But note that the consistent timestamp makes it at least possible to detect it reliably independent if the local clock is wrong. Also, note that in the last update at 10:30, the relevant (local) timestamp would have been rather 9:17 (not 10:30) because that was the last time went the client sync successfully. It's like Git commits: it matters from where you forked. And counters are a safer way to pinpoint it than using local clock time (like it is safer to say, I forked from commit XYZ than saying I forked at 10 PM yesterday).

Reply:
In my implementation, the revision timestamp is updated locally (regardless of the connection with the server). 

In your example, Browser A could update after 10:18 (offline), so it still has t=1, but the update is "the latest". Then when online, it will be overwritten by the server at 10:30 (from your example).

As I think there is no sense (in that flat few options changing rarely) to creating an event system with a log of all changes (more or less a git system) - **I strongly opt-in to make the newest (in time) change as a source of truth** - without a time, you cannot do it. A counter mechanism is not time aware.

Reply:
> With that a counter cannot be used to decide the order of events 

And that's the opposite of my idea - the "newest" in time change should have priority, as it is the most expected from the user.

> With that a counter cannot be used to decide the order of events (which happened before the other), I disagree.

Please explain how it would work with the example, I showed above (that the last change is lost with the counter).

The only drawback of a timestamp is that it might be wrong in a super edge case, which in my opinion is negligible. This is not an API, which syncs every second or minute.. as users rather set the extension and forget about it. Even if we take into account passed domains (in the future).

So.. if we agree that timestamp is fine, we have an advantage, in that it can be used for what I empathized:

>  I strongly opt-in to make the newest (in time) change as a source of truth

If you look closely at the implementation, it is even smarter, as if a server has a newer version than a local one, which is tried to push to the server, we only overwrite changed keys from the local version to the server (as currently sent local version is always "the newest"), so indeed we have some merge mechanism, if there is a conflict.

Reply:
I think, we are mixed here three things that play together but are separate concerns:
* Timestamp (logical stamps vs local clock)
* Conflict detection: how to detect a conflict
* Merging: an attempt to automatically resolve a conflict

I was originally only commenting that for conflict detection, a logical timestamp would be sufficient. Since it is only used to decide what "happened-before". But I think, the crucial part of the approaches is rather who is the source of truth and how merging is handled.

> > With that a counter cannot be used to decide the order of events
> 
> And that's the opposite of my idea - the "newest" in time change should have priority, as it is the most expected from the user.

That is now conflict resolution. The idea is to merge on the client, I understand. It is possible to do that, though my intuition would have that the server would be the only source of truth.

I haven't though a lot about merging to be honest. I was rather assuming to sync changes like most REST Apis do (fetch from the server, update locally, and send the updated state back). On conflicts, you are out of luck and have to throw them away.

Typically you have internet connection and can sync quickly after each change of a sync property, which hopefully makes conflicts unlikely enough that you can get away with it. For instance, before flipping a synced pref, first fetch the server state to be in sync. Then modify the pref and send the whole state back it to the server. (Some REST APIs try to improve by providing PATCH operations to require only one request. But they are harder to implement on the server.)

Merging on client side is possible. You end up with an almost p2p-like system, where clients manage the state and only use the server to share their state with the other clients. There is no well-defined leader and the last client that sent defines what a new extension installations would see.

If it works, it is nice. I find it only a bit hard to reason about; but maybe it works well for our use case.

(For instance, for me it is not trivial for me to see that if  multiple extensions are being logged in that the system is eventually consistent. That means, after waiting long enough - not changing anything and all are able to reach the server - if they would always end up with the same state of the sync parameters. If the database of the server is the only source of truth, it's easier to see that updates will eventually reach all clients.

Also, error recovery. If clients make changes in their local databases but fail to sync, will they eventually resolve to the same state once they can reach the server again? Maybe yes, but for me, it's not trivial to reason about that either.)

Reply:
We need to face it, that the only problem occurs with updates to the options, which both try to sync, but fail. Timestamp gives information about, which is "latest", so we can assume that this is what the user wants. 

Think of that... we are talking about a single user with multiple devices (is rather rare that it would be browsers on the same device). How often that it might be an imagined situation that we try to figure out? If a user tried to update options on mobile, but it failed and looks at the computer and sees no change, it will update it there, so then he thinks that this is the source of truth. Nothing complicated. 

About the source of truth, it is indeed the server. If it has a newer revision it overwrites the local version. And whenever options are get first check if the server has a newer version.

Reply:
Then why not keep it like the old extension did it? It fetches state from the server, overwrites its local state, and on each change, updates the server state (and retry a couple of time if it fails).

To me that would be the straightforward solution under the assumptions.

For instance, flipping a pref that is synced:
* fetch the state from the server
* update it locally (flip the pref)
* send the new state to the server

From time to time poll for changes.

(BTW, I find multiple extension running on the same device not unrealistic. For instance, people may use Firefox and Chrome with Ghostery on the same machine. I'm not aware of bug reports for the old extension regarding the syncing.)

Reply:
The state is empty on the first hit on the server, and you may not make any changes to options after logging in. The idea to fetch and save data from the server is technically fine, but I wanted to avoid setting local options when it is not required - then we have to be careful to avoid a loop of setting the local version and fetching it from the server. That's why I introduced revision. It is just like you wrote but with a check of the time of the last update.

We have a limited list of options, which only a part could be migrated, so when the user will update extension MV2 to MV3 options will be migrated locally, so when the user will be logged in **For the first time** it will send the current state to the new API end-point. It means, that migration is covered.